### ISSUE TITLE
attempt to get length of local 'addr' (a nil value)

### ISSUE THREAD

**Issue by seifchen** [Kong/kong-plugin-proxy-cache#17]
I use this plugin to cache resposne,but got nil err:
```
2020/01/04 14:43:06 [debug] 61#0: *8399 [lua] init.lua:227: [cluster_events] polling events from: 1578118130.488 to: 1578120186.791
2020/01/04 14:43:08 [debug] 61#0: *8406 [lua] init.lua:852: balancer(): setting address (try 1): nil:80
2020/01/04 14:43:08 [error] 61#0: *8406 failed to run balancer_by_lua*: /usr/local/openresty/lualib/ngx/balancer.lua:131: attempt to get length of local 'addr' (a nil value)
stack traceback:
        /usr/local/openresty/lualib/ngx/balancer.lua:131: in function 'set_current_peer'
        /usr/local/share/lua/5.1/kong/init.lua:854: in function 'balancer'
        balancer_by_lua:2: in main chunk while connecting to upstream, client: 10.17.250.102, server: kong, request: "$request"
```
but when I delete this plugin, everything is ok!

kong version: 1.4.0
here is my config:
```
{
    "created_at": 1578118040,
    "config": {
        "request_method": [
            "GET"
        ],
        "cache_control": false,
        "vary_query_params": [
            "page",
            "city"
        ],
        "storage_ttl": null,
        "response_code": [
            200
        ],
        "cache_ttl": 300,
        "vary_headers": null,
        "strategy": "memory",
        "content_type": [
            "text/plain",
            "application/json;charset=UTF-8",
            "application/json"
        ],
        "memory": {
            "dictionary_name": "kong_db_cache"
        }
    },
    "id": "431aab0d-285e-4816-8af2-95e6872f9521",
    "service": null,
    "name": "proxy-cache",
    "protocols": [
        "grpc",
        "grpcs",
        "http",
        "https"
    ],
    "enabled": true,
    "run_on": "first",
    "consumer": null,
    "route": {
        "id": "22b75294-46e7-4499-9893-345cc5d967f6"
    },
    "tags": null
}
```

**Comment by seifchen** [comment]
I found I chang the send_response method:https://github.com/Kong/kong-plugin-proxy-cache/blob/master/kong/plugins/proxy-cache/handler.lua#L245
use kong.response.exit(200,res.body) ,then the error fix

**Comment by locao** [comment]
Hi @seifchen, thanks for your report!

I tried to reproduce this issue and was not able. Can you share other details on the configuration you are using?

This log entry seems relevant: `balancer(): setting address (try 1): nil:80`. The address should not be `nil` here. How are your upstreams configured?

**Comment by seifchen** [comment]
I find the reason is I use the domain name to request this route the error happend
but use the ip eveything is ok,  there is nothing different， the domain name config by nginx:
```
upstream  test{ 
    server  $host1:80;
    server $host2:80;
}

server {
    listen 80;
    listen 443 ssl;

    ssl_session_timeout  5m;
    ssl_protocols  TLSv1 TLSv1.1 TLSv1.2;
    ssl_ciphers  HIGH:!RC4:!MD5:!aNULL:!eNULL:!NULL:!DH:!EDH:!EXP:+MEDIUM;
    ssl_prefer_server_ciphers   on;
    
    server_name $serviceName;

    location / {
        proxy_pass  http://test;
        include  custom_proxy_header.conf;
    }
    access_log  /data/service_logs/nginx/native-maodou_access.log  misc;
    error_log   /data/service_logs/nginx/native-maodou_error.log;
}
```

and I change do not use the  delayed_response_callback use this code is ok
```
ngx.ctx.delayed_response = {
   status_code = 200,
   content_type=header,
   content = res.body,
  }
```

**Comment by seifchen** [comment]
This is the step to reduce:
### Steps To Reproduce
1.  use docker to run nginx
```
docker run --name nginx -d \
-v $Path/nginx/nginx.conf:/etc/nginx/nginx.conf:ro \
-v $Path/nginx/default.conf:/etc/nginx/conf.d/default.conf:ro \
-p 9099:9099 \
nginx
```
nginx.conf is:
```
user  nginx;
worker_processes  1;
error_log  /var/log/nginx/error.log warn;
pid        /var/run/nginx.pid;
events {
    worker_connections  1024;
}
http {
    include       /etc/nginx/mime.types;
    default_type  application/octet-stream;
    sendfile        on;
    #tcp_nopush     on;
    keepalive_timeout  65;
    #gzip  on;
    include /etc/nginx/conf.d/*.conf;
}
```

default.conf is:
```
server {
    listen       9099;
    location / {
      
          proxy_pass http://10.115.0.11:8000;
          #proxy_http_version 1.1; 
    }
}
```

2. then use the curl to request the route which enable the proxy-cache,
first request will be ok ,but second request which cache hit will err
```
curl localhost:9099/tools/test/$uri
```
3. change the default.conf like this:
```
server {
    listen       9099;
    location / {
      
          proxy_pass http://10.115.0.11:8000;
          proxy_http_version 1.1; 
    }
}
```

```
docker exec nginx nginx -s reload
```

4. than use the step 2 to request everything will be ok

5. add ngx.exit(ngx.HTTP_OK) in delayed_response_callback
```
ngx.ctx.delayed_response_callback = function()
    ngx.say(res.body)
    ngx.exit(ngx.HTTP_OK)
  end
```
6. edit the default.conf enable proxy_http_version 1.1 or disable, everything will be ok

7. I write a test plugin fisrt like this:
```
local ngx = ngx
local TestHandler = {}

TestHandler.PRIORITY = 995


function TestHandler:access(config)
    ngx.ctx.delayed_response = true
    ngx.ctx.delayed_response_callback = function()
        ngx.say("test plugin")
    end
    return
end
return TestHandler
```
second like this
```
local ngx = ngx
local TestHandler = {}

TestHandler.PRIORITY = 995


function TestHandler:access(config)
    ngx.ctx.delayed_response = true
    ngx.ctx.delayed_response_callback = function()
        ngx.say("test plugin")
        ngx.exit(ngx.HTTP_OK)
    end
    return
end

return TestHandler

```

the first also happend this error and the second will be ok,so should add ngx.exit in the delayed_response_callback.

**Comment by locao** [comment]
Hi @seifchen, thanks for adding more details.

What happens if you point your request straight to Kong, like this: `$ curl localhost:8000/tools/test/$uri`?

You said that everything works as expected when you use the route's IP, so you probably have a problem with your DNS configuration, you might want to take a look at NGINX's docs, especially the `resolver` directive.

Regarding your changes to the plugin, they seem to work but what is happening is that the request is terminated when you call `ngx.exit`, so the configuration errors that may exist in the next phases are not reached.

**Comment by seifchen** [comment]
@locao I thind this plugin will avoid Kong.balancer to run, but when I not use ngx.exit the Kong.balancer exec.    use this url like this: $ curl localhost:8000/tools/test/$uri? the Kong.balancer not access, but throw nginx to proxy in front of kong ,the Kong.balancer access

**Comment by bungle** [comment]
@seifchen as we merger #19 and released `1.3.0`, I closed this. Please reopen if you think we sitll have a problem.

**Comment by u-ashish** [comment]
@bungle I can confirm upgrading plugin to `1.3` worked for me (I saw the exact same issue as OP).

It would be nice if the default bundled version in `kong` already came with `1.3` (it's on `1.2.3`). I had to manually upgrade by doing a `luarocks install kong-proxy-cache-plugin` before my `kong start` in the dockerized container.

Thanks for fixing it though! I would have been miserable if it wasn't resolved already :)
